libs/common-types/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+	TrancheId: PartialEq + PartialOrd,
+{
+	type Property = PoolRole<Moment, TrancheId>;
+
+	fn exists(&self, property: Self::Property) -> bool {
+		match property {
+			PoolRole::Borrower => self.admin.contains(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.contains(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.contains(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.contains(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.contains(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.contains(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, _) => self.tranche_investor.contains(id),
+		}
+	}
+
+	fn empty(&self) -> bool {
+		self.admin.is_empty() && self.tranche_investor.is_empty()
+	}
+
+	fn rm(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.remove(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.remove(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.remove(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.remove(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.remove(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.remove(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.remove(id, delta),
+		}
+	}
+
+	fn add(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.insert(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.insert(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.insert(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.insert(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.insert(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+		}
+	}
+}
+
+impl<Now, MaxTranches, MinDelay, TrancheId, Moment>
+	TrancheInvestors<Now, MaxTranches, MinDelay, TrancheId, Moment>
+where
+	MaxTranches: Get<TrancheId>,
+	Now: Time<Moment = Moment>,
+	MinDelay: Get<Moment>,
+	Moment: PartialEq + PartialOrd + Saturating + Ord + Copy,
+	TrancheId: PartialEq + PartialOrd,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_empty()
+	}
+
+	fn validity(&self, delta: Moment) -> Moment {
+		let now = Now::now();
+		let min_validity = now.saturating_add(MinDelay::get());
+		let req_validity = now.saturating_add(delta);
+		max(min_validity, req_validity)
+	}
+
+	pub fn contains(&self, tranche: TrancheId) -> bool {
+		if tranche >= self.max_tranches {
+			return false;
+		}
+
+		self.info
+			.iter()
+			.position(|info| info.tranche_id == tranche && info.permissioned_till >= Now::now())
+			.is_some()
+	}
+
+	pub fn remove(&mut self, tranche: TrancheId, delta: Moment) {
```

Comment:
Do we need both insert and remove? I think it could just be `set/update`, and just always do the `MinDelay` check.

Reply:
mmh. Probably not. Its actually already doing what you are mentioning. Do you want this to be renamed?